mongomock/aggregate.py

```diff
@@ -622,19 +618,35 @@ def _handle_project_stage(in_collection, unused_database, options):
             raise ValueError(
                 'Bad projection specification, cannot include fields '
                 'or add computed fields during an exclusion projection: %s' % options)
-        # TODO(pascal): Do not modify input collection.
-        out_collection = _extend_collection(in_collection, field, value)
-        if field != '_id':
+        if field == '_id':
+            continue
+        if value in (0, 1, True, False):
             filter_list.append(field)
+            continue
+        if not new_fields_collection:
+            new_fields_collection = [{} for unused_doc in in_collection]
+        _extend_collection(in_collection, new_fields_collection, field, value)
     if (method == 'include') == (include_id is not False):
         filter_list.append('_id')
-    return [
+
+    if not filter_list:
+        return new_fields_collection
+
+    # Final steps: include or exclude fields and merge with
+    # newly created fields.
+    out_collection = [
         {
             k: v for (k, v) in x.items()
             if (method == 'include') == (k in filter_list)
         }
-        for x in out_collection
+        for x in in_collection
     ]
+    if new_fields_collection:
+        return [
+            dict(a, **b)
```

Comment:
what happens here if there are some common keys between the two dicts? Please, make sure the behaviour is consistent with MongoDB doc, and add test to make sure of it.

Reply:
It's actually not possible: fields in options can either be added to new_field_collections xor to filter_list (as it's added only once), and in a case of exclusion, new_field_collections is empty so there cannot be a conflict.

Thanks for asking though, I've added a test with an exclusion and a computed field (that raises an error).